session/artifactmap.go

```diff
@@ -12,175 +12,191 @@ import (
+	ArtifactEntry struct {
+		Name         string      `json:"name"`
+		ID           string      `json:"resourceId"`
+		ItemID       string      `json:"ID,omitempty"`
+		ResourceType string      `json:"resourceType"`
+		Data         interface{} `json:"-"`
 	}
 
-	// AppData struct to unmarshal list app response
-	AppData struct {
-		Data []AppsResp `json:"data"`
+	// ItemData struct to unmarshal list app response
+	ItemData struct {
+		Data []ArtifactEntry `json:"data"`
 	}
 
-	ArtifactListDict []*ArtifactEntry
+	// ArtifactList all items, spaces and collections
+	ArtifactList struct {
+		list   []*ArtifactEntry
+		sorted bool
+	}
 
 	// ArtifactMap is the map between app names and GUIDs
 	ArtifactMap struct {
-		appTitleToID     *sync.Map
-		appTitleToItemID *sync.Map
-		streamTitleToID  *sync.Map
-		spaceTitleToID   *sync.Map
-		AppList          ArtifactListDict
-		mu               sync.Mutex
+		resourceMap map[string]*ArtifactList
+		mu          sync.RWMutex
 	}
 
-	// ArtifactEntry is a key value pair
-	ArtifactEntry struct {
-		Title  string
-		GUID   string
-		ItemID string
+	// ArtifactKey used to find item using title
+	ArtifactKey struct {
+		Title string
+		Type  string
 	}
 
+	// SpaceNameNotFoundError space with name not found in artifact map error
 	SpaceNameNotFoundError string
-	SpaceIdNotFoundError   string
+	// SpaceIDNotFoundError space with ID not found in artifact map error
+	SpaceIDNotFoundError string
+)
+
+// Currently known resource types
+const (
+	ResourceTypeApp         = "app"
+	ResourceTypeGenericLink = "genericlink"
+	ResourceTypeDataset     = "dataset"
+	ResourceTypeDataAsset   = "dataasset"
+
+	// none Sense "resource types"
+	ResourceTypeStream = "stream"
+	ResourceTypeSpace  = "space"
 )
 
 // Error implements error interface
 func (err SpaceNameNotFoundError) Error() string {
 	return fmt.Sprintf("space name <%s> not found", string(err))
 }
 
-func (err SpaceIdNotFoundError) Error() string {
+func (err SpaceIDNotFoundError) Error() string {
 	return fmt.Sprintf("space id <%s> not found", string(err))
 }
 
-func (d ArtifactListDict) Len() int {
-	return len(d)
+func (d *ArtifactList) Len() int {
+	if d == nil {
+		return 0
+	}
+	return len(d.list)
 }
-func (d ArtifactListDict) Less(i, j int) bool {
-	return d[i].Title < d[j].Title
+
+func (d ArtifactList) Less(i, j int) bool {
+	return d.list[i].Name < d.list[j].Name
 }
-func (d ArtifactListDict) Swap(i, j int) {
-	d[i], d[j] = d[j], d[i]
+
+func (d ArtifactList) Swap(i, j int) {
+	d.list[i], d.list[j] = d.list[j], d.list[i]
 }
 
 // NewAppMap returns an empty ArtifactMap
 func NewAppMap() *ArtifactMap {
-	return &ArtifactMap{
-		appTitleToID:     &sync.Map{},
-		streamTitleToID:  &sync.Map{},
-		appTitleToItemID: &sync.Map{},
-		spaceTitleToID:   &sync.Map{},
-	}
+	return &ArtifactMap{resourceMap: make(map[string]*ArtifactList)}
 }
 
-// IsEmpty returns true of entry is nil or has no GUID
-func (entry *ArtifactEntry) IsEmpty() bool {
+// Copy of ArtifactEntry
```

Comment:
Data is not copied. Add this info to doc-comment if this is on purpose.

Reply:
no, just missed than one when adding .Data, will fix